altapay.php

```diff
@@ -0,0 +1,2353 @@
+     * @param $vouchers
+     * @param $productID
+     * @param $discountPercent
+     * @param $basePrice
+     * @param $orderSubtotal
+     * @param $freeGiftVoucher
+     * @return float|int
+     */
+    private function getVoucherDiscounts($vouchers, $productID, $discountPercent, $basePrice, $orderSubtotal, $freeGiftVoucher)
+    {
+        $discountedAmount = 0;
+        $productPriceAfterDiscount = 0;
+        foreach ($vouchers as $key => $voucher) {
+            if (in_array($productID, $voucher['products']) || $voucher['products'] == 'all') {
+                if (empty($discountPercent) && $voucher['reductionPercent'] != '0.00') {
+                    $discountPercent += $voucher['reductionPercent'];
+                    $discountedAmount = $basePrice * ($discountPercent / 100);
+                    $productPriceAfterDiscount = $basePrice - $discountedAmount;
+                } else if ($voucher['reductionPercent'] == '0.00' && (empty($freeGiftVoucher['free_gift']) || $freeGiftVoucher['free_gift'] )&& $freeGiftVoucher['free_gift'] != $productID) {
+                    if($freeGiftVoucher['free_gift']) {
+                        $discountPercent += (($freeGiftVoucher['reductionAmount']+$freeGiftVoucher[$key])/($orderSubtotal+$freeGiftVoucher[$key])*100);
+                    }
+                    else {
+                        $discountPercent += ($freeGiftVoucher[$key]/ ($orderSubtotal)) * 100;
+                    }
+
+                } else if ($voucher['reductionPercent'] == '0.00' && empty($freeGiftVoucher['free_gift']) || $freeGiftVoucher['free_gift'] == $productID) {
+                    $discountPercent += (($freeGiftVoucher['reductionAmount']+$freeGiftVoucher[$key])/($orderSubtotal+$freeGiftVoucher[$key])*100);
+
+                } else {
+                    $totalDiscountedAmount = $discountedAmount + ($productPriceAfterDiscount * ($voucher['reductionPercent'] / 100));
+                    $discountPercent = ($totalDiscountedAmount / $basePrice) * 100;
+                }
+            }
+
+        }
+
+        return $discountPercent;
+    }
+
+
+    /**
+     * Returns the array of Cart Rule properties like discount percentage, free shipping status and free gift conditions
+     * @param $cart
+     * @return array
+     */
+    private function getCartRuleProperties($cart)
+    {
+        $j = 0;
+        $voucherProperties = array();
+        $freeShipping = array();
+        $cartRules = $cart->getCartRules();
+
+        foreach ($cartRules as $key => $cartRule) {
+            if ($cartRule['gift_product']) {
+                $voucherProperties['free_gift'] = $cartRule['gift_product'];
+            }
+            $cartRuleID = $cartRule['id_cart_rule'];
+            $freeShipping[$j] = $cartRule['free_shipping'];
+            $voucherProperties[$cartRuleID] = $cartRule['value_real'];
+            $voucherProperties['reductionAmount'] = $cartRule['reduction_amount'];
+            $j++;
```

Comment:
Simply use auto increment:
```suggestion
        $voucherProperties = array();
        $freeShipping = array();
        $cartRules = $cart->getCartRules();

        foreach ($cartRules as $key => $cartRule) {
            if ($cartRule['gift_product']) {
                $voucherProperties['free_gift'] = $cartRule['gift_product'];
            }
            $cartRuleID = $cartRule['id_cart_rule'];
            $freeShipping[] = $cartRule['free_shipping'];
            $voucherProperties[$cartRuleID] = $cartRule['value_real'];
            $voucherProperties['reductionAmount'] = $cartRule['reduction_amount'];
```